Add config file documentation

senofsky opened the pull request:
Fixes #78 #92 

This pull requests add brief documentation for using configuration files.

I tried to keep it brief for quick usability purposes, but I'd love to get feedback for any improvements.

I would like to note that markdown does not support embedding files into documents, which made it difficult to embed an existing configuration file into the documentation. My final solution was, non-ideally, to copy/paste the configuration file into the documentation.

However, I was able to embed the configuration file directly into the document by using [mkdocs macros](https://github.com/fralau/mkdocs_macros_plugin) and creating the following code snippet:

```python
import html                                                                                             
import os                                                                                               
                                                                                                                                                                                                      
def define_env(env):                                                                                    
                                                                                                        
    @env.macro                                                                                          
    def load_code_from_file(file_path: str, flavor: str = ""):                                          
       """                                                                                             
       Load code from a file and save as a preformatted code block.                                    
       If a flavor is specified, it's passed in as a hint for syntax highlighters.                     
                                                                                                        
       Example usage in markdown:                                                                      
                                                                                                        
          {{load_code_from_file("code/myfile.py", "python")}}                                         
                                                                                                        
       """                                                                                             
       docs_dir = env.variables.get("docs_dir", "docs")                                                
       file_path = os.path.abspath(os.path.join(docs_dir, file_path))                                  
       if not os.path.exists(file_path):                                                               
          return f"""<b>File not found: {file_path}</b>"""                                            
       with open(file_path, "r") as file:                                                              
          return (                                                                                    
             f"""<pre><code class="{flavor}">{html.escape(file.read())}</code></pre>"""              
          )   
```

but the `mkdocs-macros-plugin` requires this macro to be located in the project root directory in a module named `main.py`. To me, having a `main.py` in the project root would be ugly and confusing, so I chose not to use this solution.

senofsky committed:
Renamed exhaustive-config to comprehensive-config

This new name better expresses the idea that this configuration file
includes all possible options

senofsky committed:
Add documentation for using configuration files

senofsky committed:
Update comprehensive-config with new values

ALSchwalm commented on docs/configuration-file/comprehensive-example.md:
```diff
@@ -0,0 +1,71 @@
+ssh-bin-name     = "/sbin/ssh"
+ssh-command      = "whoami"
+ssh-console      = false
+ssh-port         = 22
```
This should probably be a high port, just so it is clear to people that 22 is _not_ the default.

senofsky replied:
From https://transient.readthedocs.io/en/latest/cli/run/:
`-ssh-port PORT: Connect to port PORT on the virtual machine for SSH instead of the default 22.`

Or am I grossly misunderstanding something?

ALSchwalm commented on test/config-files/comprehensive-config:
```diff
@@ -34,18 +34,18 @@ copy-in-before   = []
+ssh-bin-name     = "/sbin/ssh"
+ssh-command      = "whoami"
 ssh-console      = false
 ssh-port         = 22
```
Same thing here about the port

senofsky replied:
From https://transient.readthedocs.io/en/latest/cli/run/:
`-ssh-port PORT: Connect to port PORT on the virtual machine for SSH instead of the default 22.`

Or am I grossly misunderstanding something?

ALSchwalm commented on test/config-files/comprehensive-config:
```diff
@@ -34,18 +34,18 @@ copy-in-before   = []
-# image_backend    = None
-# image_frontend   = None
-# name             = None
+image_backend    = "$HOME/.local/share/transient/backend"
```
Does the variable expansion here actually work?

ALSchwalm replied:
Oh, I see you opened #92. If we're going to do that, then this PR is blocked until we add the variable expansion feature.

senofsky replied:
It does not, but I opened https://github.com/ALSchwalm/transient/issues/92 because of this.

Maybe the best solution would be for me to change the path and add a comment that indicates variable expansion is not currently supported. Then we can remove it after we add that feature.

senofsky replied:
Added commit that supports environment variable expansion and mentioned the fix in the PR body.

ALSchwalm reviewed (changes requested):
Weird about the macros thing. Seems like there must be a way to work around that, but this approach should be fine for now.

ALSchwalm commented:
Can you mark this as fixes #78

senofsky committed:
Add support for using env variables in config file

ALSchwalm commented on docs/configuration-file/comprehensive-example.md:
```diff
@@ -0,0 +1,71 @@
+### Comprehensive Example
```
I appreciate that we can't make this pull in the file automatically (without the macro hacks you mentioned), but I think I would still at least like to generate this file via a make rule. So maybe the `make docs` will build this before serving so we have a (better) chance of this not getting out-of-sync.

ALSchwalm reviewed (changes requested):
Looks good to me, just one comment

senofsky committed:
Update "make docs" to generate doc for config file

The comprehensive configuration file displayed in the documentation
should be the same one being tested. Therefore, the configuration file
is embedded into the documentation:
- Add the template docs/template/comprehensive-example-template.md
- Add the script embed_file_into_markdown_template.py
- Update the Makefile target "docs" to generate comprehensive-example.md
- Update .gitignore to ignore generated documents
- Add docs/templates/README.md to provide instructions on using templates
